### PR TITLE
Test_autograd support for 0-dim input/outputs.

### PR DESCRIPTION
**Pull request by gchanan**
This uses the 'fake' _scalar_sum function to test scalar (0-dimensional) inputs and output in test_autograd.

Main changes:
1) Introduces a randn_like function (this is really just for convience but it comes up often in testing.
2) Because the Tensor and Variable API are different wrt sizes, we take care to not exit the Variable API when constructing Variables based on other Variables.  This is pretty straightforward, but there is sometimes an extra line of code for setting requires_grad.  Should we have the 'like' functions maintain requires_grad?  Or bind all factory functions with an additional 'requires_grad' parameter?

### CODE REVIEW

**Comment by gchanan** [comment]
This builds on the _scalar_sum function which will go away when we have full 0-dim tensor support; if there's an objection to having this temporarily I can wait to commit, but I wanted to have a fairly standalone PR for this.

**Comment by gchanan** [comment]
actually, I'm going to get rid of the explicit _scalar_sum tests (but leave _scalar_sum for easier local testing); they are a bit too fragile right now to keep.

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def get_analytical_jacobian(input, output):
     input = contiguous(input)
     jacobian = make_jacobian(input, output.numel())
     jacobian_reentrant = make_jacobian(input, output.numel())
-    grad_output = output.data.clone().zero_()
+    grad_output = output.clone().detach_().zero_()
```
Comment (colesbury) [inline torch/autograd/gradcheck.py]: `grad_output = torch.zeros_like(output)` ?

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -193,7 +193,7 @@ def fn(input):
     zero_gradients(inputs)
     output = _differentiable_outputs(func(*inputs))
     if any([o.requires_grad for o in output]):
-        torch.autograd.backward(output, [o.data.new(o.size()).zero_() for o in output], create_graph=True)
+        torch.autograd.backward(output, [o.new(o.size()).zero_() for o in output], create_graph=True)
```
Comment (colesbury) [inline torch/autograd/gradcheck.py]: `torch.zeros_like(o)`?

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -635,6 +635,9 @@
 - name: _standard_gamma_grad(Tensor self, Tensor output)
   self: not_implemented("_standard_gamma_grad")
 
+- name: _scalar_sum(Tensor self)
+  self: _scalar_sum_backward(grad, self)
```
Comment (colesbury) [inline tools/autograd/derivatives.yaml]: `grad.expand(self.sizes())`

Don't make people jump to another file to read such a simple gradient definition.

**Review by colesbury — COMMENTED** [review]
I think `randn_like` should be added to `DONT_REQUIRE_DERIVATIVE` in `gen_variable_type.py`

I'm pretty ambivalent about adding it as a new factory method. Not really sure it's useful outside of the gradient checking code

File: tools/autograd/derivatives.yaml (hunk 3 above)
Comment (gchanan) [reply tools/autograd/derivatives.yaml]: hehe, this actually doesn't work and one of the reasons I disabled the test for being too fragile.  The reason is we change the size of things inplace (to get around the scalar blocks) so you actually need the tensor to get the right size, not the saved sizes of tensor.

**Comment by gchanan** [comment]
Ok, I used `zeros_like` where it was requested, added `randn_like` to DONT_REQUIRE_DERIVATIVE, and got rid of the _scalar_sum binding.

**Review by colesbury — APPROVED** [review]
Still not really sure if we should permanently support `randn_like`. If we do, we should document it.
